File: plugins/plugin-im/model/im_file.go
package model

import "time"

// ImFile stores files uploaded through the IM module.
// Reuses the same storage engine as SysFile but uses a dedicated table.
type ImFile struct {
	ID           string `gorm:"primaryKey;size:32" json:"id"`
	Engine       string `gorm:"size:32;not null" json:"engine"`                 // LOCAL, MINIO, S3
	Bucket       string `gorm:"size:128;not null" json:"bucket"`                // storage bucket
	FileKey      string `gorm:"size:500;not null;index" json:"file_key"`        // unique object key in storage
	Name         string `gorm:"size:255;not null" json:"name"`                  // original filename
	Suffix       string `gorm:"size:32" json:"suffix"`                          // file extension
	SizeKb       int64  `gorm:"default:0" json:"size_kb"`                       // file size in KB
	SizeInfo     string `gorm:"size:32" json:"size_info"`                       // formatted size
	StoragePath  string `gorm:"size:500" json:"storage_path"`                   // path in storage backend
	DownloadPath string `gorm:"size:500" json:"download_path"`                  // HTTP download URL
	Thumbnail    string `gorm:"size:500" json:"thumbnail"`                      // thumbnail URL for images
	Checksum     string `gorm:"size:128" json:"checksum"`                       // SHA256 hex
	ChecksumAlgo string `gorm:"size:16" json:"checksum_algo"`                   // "sha256"

	// IM-specific fields
	ConversationID string `gorm:"size:32;index" json:"conversation_id"`
	SenderID       string `gorm:"size:32;not null;index" json:"sender_id"`
	SenderType     string `gorm:"size:20;not null" json:"sender_type"`
	MsgType        string `gorm:"size:20;not null" json:"msg_type"` // IMAGE | FILE

	CreatedAt time.Time `json:"created_at"`
}

func (ImFile) TableName() string { return "im_file" }

// ImFileVO is the view object for IM file responses.
type ImFileVO struct {
	ID           string `json:"id"`
	Engine       string `json:"engine"`
	Bucket       string `json:"bucket"`
	FileKey      string `json:"file_key"`
	Name         string `json:"name"`
	Suffix       string `json:"suffix"`
	SizeKb       int64  `json:"size_kb"`
	SizeInfo     string `json:"size_info"`
	DownloadPath string `json:"download_path"`
	Thumbnail    string `json:"thumbnail"`

	ConversationID string `json:"conversation_id"`
	SenderID       string `json:"sender_id"`
	SenderType     string `json:"sender_type"`
	MsgType        string `json:"msg_type"`

	CreatedAt string `json:"created_at"`
}

func (e *ImFile) ToVO() *ImFileVO {
	createdAt := ""
	if !e.CreatedAt.IsZero() {
		createdAt = e.CreatedAt.Format("2006-01-02 15:04:05")
	}
	return &ImFileVO{
		ID:             e.ID,
		Engine:         e.Engine,
		Bucket:         e.Bucket,
		FileKey:        e.FileKey,
		Name:           e.Name,
		Suffix:         e.Suffix,
		SizeKb:         e.SizeKb,
		SizeInfo:       e.SizeInfo,
		DownloadPath:   e.DownloadPath,
		Thumbnail:      e.Thumbnail,
		ConversationID: e.ConversationID,
		SenderID:       e.SenderID,
		SenderType:     e.SenderType,
		MsgType:        e.MsgType,
		CreatedAt:      createdAt,
	}
}

File: plugins/plugin_im/model/im_file.py
"""IM file upload ORM model."""

from __future__ import annotations

from datetime import datetime
from sqlalchemy import Index, text
from sqlalchemy.dialects.mysql import BIGINT, DATETIME, VARCHAR
from sqlalchemy.orm import Mapped, mapped_column
from sdk.kernel.registry import HeiBase


class ImFile(HeiBase):
    """IM模块上传文件记录"""
    __tablename__ = "im_file"
    __table_args__ = (
        Index("idx_file_key", "file_key"),
        Index("idx_conversation_id", "conversation_id"),
        Index("idx_sender_id", "sender_id"),
    )

    id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), primary_key=True)
    engine: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    bucket: Mapped[str] = mapped_column(VARCHAR(128, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    file_key: Mapped[str] = mapped_column(VARCHAR(500, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    name: Mapped[str] = mapped_column(VARCHAR(255, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    suffix: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    size_kb: Mapped[int] = mapped_column(BIGINT, default=0, server_default=text("0"))
    size_info: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    storage_path: Mapped[str] = mapped_column(VARCHAR(500, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    download_path: Mapped[str] = mapped_column(VARCHAR(500, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    thumbnail: Mapped[str] = mapped_column(VARCHAR(500, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    checksum: Mapped[str] = mapped_column(VARCHAR(128, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    checksum_algo: Mapped[str] = mapped_column(VARCHAR(16, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    conversation_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    sender_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    sender_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    msg_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DATETIME, nullable=False)
